cpp/arcticdb/version/version_map_batch_methods.hpp

```diff
@@ -117,6 +120,132 @@ inline std::shared_ptr<std::unordered_map<std::pair<StreamId, VersionId>, AtomKe
+       },
+    [] (const auto&) -> std::optional<AtomKey> {
+        util::raise_rte("Unsupported version query type");
+    });
+}
+
+inline std::vector<folly::Future<std::optional<AtomKey>>> batch_get_versions(
+    const std::shared_ptr<Store>& store,
+    const std::shared_ptr<VersionMap>& version_map,
+    const std::vector<StreamId>& symbols,
+    const std::vector<pipelines::VersionQuery>& version_queries) {
+    ARCTICDB_SAMPLE(BatchGetVersion, 0)
+    util::check(symbols.size() == version_queries.size(), "Symbol and version query list mismatch: {} != {}", symbols.size(), version_queries.size());
+
+    robin_hood::unordered_flat_map<StreamId, StreamVersionData> version_data;
+    for(const auto& symbol : folly::enumerate(symbols)) {
+        auto it = version_data.find(*symbol);
+        if(it == version_data.end())
+            version_data.insert(robin_hood::pair<StreamId, StreamVersionData>(*symbol, StreamVersionData{version_queries[symbol.index]}));
+        else
+            it->second.react(version_queries[symbol.index]);
+    }
+
+    using SplitterType = folly::FutureSplitter<std::shared_ptr<VersionMapEntry>>;
+    robin_hood::unordered_flat_map<StreamId, SplitterType> shared_futures;
+    std::vector<folly::Future<std::optional<AtomKey>>> output;
+    output.reserve(symbols.size());
+    for(const auto& symbol : folly::enumerate(symbols)) {
+        const auto it = version_data.find(*symbol);
+        util::check(it != version_data.end(), "Missing version data for symbol {}", *symbol);
+        if(it->second.count_ == 1) {
+            output.push_back(async::submit_io_task(CheckReloadTask{store, version_map, *symbol,
+                                                                   it->second.load_param_}).thenValue([version_query =
+            version_queries[symbol.index]](auto version_map_entry) {
+                return get_key_for_version_query(version_map_entry, version_query);
+            }));
+        } else {
+            auto fut = shared_futures.find(*symbol);
+            if(fut == shared_futures.end()) {
+                auto [splitter, inserted] = shared_futures.emplace(*symbol, folly::FutureSplitter(async::submit_io_task(CheckReloadTask{store, version_map, *symbol,
+                                                                                                                                                                           it->second.load_param_})));
+                output.push_back(splitter->second.getFuture().thenValue([version_query = version_queries[symbol.index]](auto version_map_entry) {
+                    return get_key_for_version_query(version_map_entry, version_query);
+                }));
+            } else {
+                output.push_back(fut->second.getFuture().thenValue([version_query = version_queries[symbol.index]](auto version_map_entry) {
+                    return get_key_for_version_query(version_map_entry, version_query);
+                }));
```

Comment:
These three lines are almost identical in all 3 branches. Consider extracting the part before `.thenValue` to a variable and do the `.thenValue` outside the `if`?

Reply:
Agree. Done